File: Task4/task4.cpp
#include <iostream>
#include <vector>
using namespace std;

int bestMove(vector<int>& a, int pos, int n, int m, int lastTake) {
    int limit = min(m, n - pos);

    int bestSum = -2000000000;
    int bestK = 1;
    int curSum = 0;

    for (int k = 1; k <= limit; k++) {
        curSum += a[pos + k - 1];

        if (k == lastTake) continue;

        if (curSum > bestSum || (curSum == bestSum && k < bestK)) {
            bestSum = curSum;
            bestK = k;
        }
    }

    return bestK;
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<int> a(n);
    for (int i = 0; i < n; i++) cin >> a[i];

    int pavel = 0, vika = 0;
    int lastPavel = 0, lastVika = 0;

    int pos = 0;
    int turn = 0;

    while (pos < n) {
        int take;

        if (turn == 0)
            take = bestMove(a, pos, n, m, lastPavel);
        else
            take = bestMove(a, pos, n, m, lastVika);

        int sum = 0;
        for (int i = 0; i < take; i++)
            sum += a[pos + i];

        if (turn == 0) {
            pavel += sum;
            lastPavel = take;
        } else {
            vika += sum;
            lastVika = take;
        }

        pos += take;
        turn = 1 - turn;
    }

    cout << (pavel > vika ? 1 : 0);
    return 0;
}

File: Task4/task4.rs
use std::io;

fn best_move(a: &[i32], pos: usize, n: usize, m: usize, last_take: i32) -> usize {
    let limit = std::cmp::min(m, n - pos);
    let mut best_sum = i32::MIN;
    let mut best_k = 1;
    let mut cur_sum = 0;

    for k in 1..=limit {
        cur_sum += a[pos + k - 1];
        // Нельзя повторять размер предыдущего хода соперника
        if k as i32 == last_take {
            continue;
        }
        if cur_sum > best_sum || (cur_sum == best_sum && k < best_k) {
            best_sum = cur_sum;
            best_k = k;
        }
    }
    best_k
}

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read n, m");
    let mut iter = input.split_whitespace();
    let n: usize = iter.next().unwrap().parse().expect("Invalid n");
    let m: usize = iter.next().unwrap().parse().expect("Invalid m");

    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read array");
    let a: Vec<i32> = input
        .split_whitespace()
        .map(|x| x.parse().expect("Invalid number"))
        .collect();
    assert_eq!(a.len(), n, "Array size does not match n");

    let mut pavel = 0;
    let mut vika = 0;
    let mut last_pavel = 0;
    let mut last_vika = 0;
    let mut pos = 0;
    let mut turn = 0; // 0 – Павел, 1 – Вика

    while pos < n {
        let take = if turn == 0 {
            best_move(&a, pos, n, m, last_pavel)
        } else {
            best_move(&a, pos, n, m, last_vika)
        };

        let sum: i32 = a[pos..pos + take].iter().sum();

        if turn == 0 {
            pavel += sum;
            last_pavel = take as i32;
        } else {
            vika += sum;
            last_vika = take as i32;
        }

        pos += take;
        turn ^= 1; // смена игрока
    }

    println!("{}", if pavel > vika { 1 } else { 0 });
}